File: benchmarks/blur/java/ImageUtil.java
package benchmarks.blur.java;

import java.awt.image.BufferedImage;
import java.awt.image.Raster;
import java.io.File;
import java.io.IOException;
import javax.imageio.ImageIO;

class ImageUtil {
    /**
     * Returns a 2D integer array extracted from the given image
     * 
     * @param path Path to image, relative from the root
     * @return 2D integer array with each pixel represented with 3 numbers for RGB
     *         channels
     */
    public static int[][] loadImage(String path) {
        System.out.print("Loading image...");

        int img[][] = null;

        try {
            BufferedImage bi = ImageIO.read(new File(path));
            Raster r = bi.getData();
            int width = r.getWidth();
            int height = r.getHeight();
            int[] pixels = r.getPixels(0, 0, width, height, (int[]) null);
            img = MatrixUtil.unflattenArray(pixels, width * 3, height); // width * 3 to account for rgb channel for each
                                                                        // pixel
        } catch (IOException e) {
            System.err.println(e);
        }

        System.out.print("\r\033[K");

        return img;
    }

    /**
     * Add reflection padding to image and also convert into a 2D double array
     * 
     * @param img     2D integer array of an image
     * @param padSize Desired size of the padding
     * @return Padded image converted to a 2D double array with width and height
     *         increased by 2 * padSize
     */
    public static double[][] padImage(int[][] img, int padSize) {
        System.out.print("Padding image...");

        int height = img.length;
        int width = img[0].length;
        int paddedWidth = width + 2 * padSize * 3;
        int paddedHeight = height + 2 * padSize;

        double[][] padded = new double[paddedHeight][paddedWidth];

        // Copy original image to center
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                padded[padSize + y][padSize * 3 + x] = img[y][x];
            }
        }

        // Fill horizontal paddings using reflection (exclude corners)
        for (int y = 0; y < padSize; y++) {
            for (int x = 0; x < width; x++) {
                // Top padding
                padded[y][padSize * 3 + x] = img[padSize - y][x];
                // Bottom padding
                padded[padSize + height + y][padSize * 3 + x] = img[height - 1 - y][x];
            }
        }

        // Fill vertical paddings (with corners)
        for (int y = 0; y < paddedHeight; y++) {
            for (int x = 0; x < padSize * 3; x += 3) {
                // Left padding
                padded[y][x] = padded[y][(padSize + padSize) * 3 - x];
                padded[y][x + 1] = padded[y][(padSize + padSize) * 3 - x + 1];
                padded[y][x + 2] = padded[y][(padSize + padSize) * 3 - x + 2];

                // Right padding
                padded[y][x + width + padSize * 3] = padded[y][width + padSize * 3 - x - 3];
                padded[y][x + width + padSize * 3 + 1] = padded[y][width + padSize * 3 - x - 2];
                padded[y][x + width + padSize * 3 + 2] = padded[y][width + padSize * 3 - x - 1];
            }
        }

        System.out.print("\r\033[K");

        return padded;
    }

    /**
     * Convert 2D double array into a jpg image
     * 
     * @param img        2D double array of an image
     * @param outputPath Desired path of the output jpg file
     */
    public static void saveImage(double[][] img, String outputPath) {
        System.out.print("Saving image...");

        int height = img.length;
        int width = img[0].length / 3;

        BufferedImage bi = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                int r = (int) Math.round(img[y][3 * x]);
                int g = (int) Math.round(img[y][3 * x + 1]);
                int b = (int) Math.round(img[y][3 * x + 2]);

                // Ensure values are in valid range
                r = Math.min(255, Math.max(0, r));
                g = Math.min(255, Math.max(0, g));
                b = Math.min(255, Math.max(0, b));

                // Combine into RGB pixel
                int rgb = (r << 16) | (g << 8) | b;

                // Set the pixel in the image
                bi.setRGB(x, y, rgb);
            }
        }

        try {
            File outputFile = new File(outputPath);
            ImageIO.write(bi, "jpg", outputFile);
        } catch (IOException e) {
            System.err.println("Error saving image: " + e.getMessage());
        }

        System.out.print("\r\033[K");
    }
}

File: benchmarks/blur/python/image_util.py
from PIL import Image
import numpy as np
from matrix_util import MatrixUtil


class ImageUtil:
    @staticmethod
    def load_image(path: str) -> list[list[int]]:
        """
        Returns a 2D integer Array extracted from the given image

        Args:
            path: Path to image, relative from the root
        Returns:
            img: 2D integer array with each pixel represented with 3 numbers for RGB channels
        """

        print("Loading image...", end="", flush=True)

        img: list[list[int]] = []

        with Image.open(path) as im:
            pixels = np.array(im).flatten()
            img = MatrixUtil.unflatten_array(
                pixels, im.width * 3, im.height)

        print("\r\033[K", end="")

        return img

    @staticmethod
    def pad_image(img: list[list[int]], pad_size: int) -> list[list[float]]:
        """
        Add reflection padding to image and also convert into a 2D float array

        Args:
            img: 2D integer array of an image
            pad_size: Desired size of the padding
        Returns:
            padded: Padded image converted to a 2D float array with width and height increased by 2 * padSize
        """

        print("Padding image...", end="", flush=True)

        height = len(img)
        width = len(img[0])
        padded_width = width + 2 * pad_size * 3
        padded_height = height + 2 * pad_size

        padded: list[list[float]] = [
            [0.0 for _ in range(padded_width)] for _ in range(padded_height)]

        # Copy original image to center
        for y in range(height):
            for x in range(width):
                padded[pad_size + y][pad_size * 3 + x] = img[y][x]

        # Fill horizontal paddings using reflection (exclude corners)
        for y in range(pad_size):
            for x in range(width):
                # Top padding
                padded[y][pad_size * 3 + x] = img[pad_size - y][x]
                # Bottom padding
                padded[pad_size + height + y][pad_size *
                                              3 + x] = img[height - 1 - y][x]

        # Fill vertical paddings (with corners)
        for y in range(padded_height):
            for x in range(0, pad_size * 3, 3):
                # Left padding
                padded[y][x] = padded[y][(pad_size + pad_size) * 3 - x]
                padded[y][x + 1] = padded[y][(pad_size + pad_size) * 3 - x + 1]
                padded[y][x + 2] = padded[y][(pad_size + pad_size) * 3 - x + 2]

                # Right padding
                padded[y][x + width + pad_size *
                          3] = padded[y][width + pad_size * 3 - x - 3]
                padded[y][x + width + pad_size * 3 +
                          1] = padded[y][width + pad_size * 3 - x - 2]
                padded[y][x + width + pad_size * 3 +
                          2] = padded[y][width + pad_size * 3 - x - 1]

        print("\r\033[K", end="")

        return padded

    def save_image(img: list[list[float]], output_path: str):
        """
        Convert 2D float array into a jpg image

        Args:
            img: 2D float array of an image
            output_path: desired path of the output jpg file
        """

        print("Saving image...", end="", flush=True)

        height = len(img)
        width = len(img[0]) // 3

        img_3d = np.array(img, dtype=np.float32).reshape((height, width, 3))
        img_3d = np.clip(img_3d, 0, 255).astype(np.uint8)

        im = Image.fromarray(img_3d, mode="RGB")
        im.save(output_path, format="JPEG")

        print("\r\033[K", end="")
